brainlit/algorithms/trace_analysis/fit_spline.py

```diff
@@ -0,0 +1,321 @@
+def checkIfDuplicates_2(listOfElems):
+    """Check if given list contains any duplicates
+    Args:
+        listOfElems (set): Build an unordered collection of unique elements.
+    """
+    setOfElems = set()
+    for elem in listOfElems:
+        if elem in setOfElems:
+            return True
+        else:
+            setOfElems.add(elem)
+    return False
+
+
+"""
+Geometric Graph class
+"""
+
+
+class GeometricGraph(nx.Graph):
+    r"""Class for undirected graphs.
+
+    It extends `nx.Graph`.
+    """
+
+    def __init__(self):
+        super(GeometricGraph, self).__init__()
+        self.segments = None
+        self.cycle = None
+        self.root = 1
+
+    def fit_spline_tree_invariant(self):
+        r"""Construct a spline tree based on the path lengths.
+
+        Raises:
+            ValueError: check if every node is unigue in location
+            ValueError: check if every node is assigned to at least one edge
+            ValueError: check if the graph contains undirected cycle(s)
+            ValueErorr: check if the graph has disconnected segment(s)
+
+        Returns:
+            spline_tree: nx.DiGraph a parent tree with the longest path in the directed graph
+        """
+
+        # check integrity of 'loc' attributes in the neuron
+        if any([self.nodes[node].get("loc") is None for node in self.nodes]):
+            raise KeyError("some nodes are missing the 'loc' attribute")
+        for node in self.nodes:
+            check_type(self.nodes[node].get("loc"), np.ndarray)
+        if any([self.nodes[node].get("loc").ndim != 1 for node in self.nodes]):
+            raise ValueError("nodes must be flat arrays")
+        if any([len(self.nodes[node].get("loc")) == 0 for node in self.nodes]):
+            raise ValueError("nodes cannot have empty 'loc' attributes")
+        for node in self.nodes:
+            check_iterable_type(self.nodes[node].get("loc"), (np.integer, np.float))
+        if any([len(self.nodes[node].get("loc")) != 3 for node in self.nodes]):
+            raise ValueError("'loc' attributes must contain 3 coordinates")
+
+        # check there are no duplicate nodes
+        LOCs = [np.ndarray.tolist(self.nodes[node]["loc"]) for node in self.nodes]
+        LOCs.sort()
+        unique_LOCs = list(LOC for LOC, _ in itertools.groupby(LOCs))
+        if len(LOCs) != len(unique_LOCs):
+            raise ValueError("there are duplicate nodes")
+
+        # check the graph is edge-covering
+        if not nx.algorithms.is_edge_cover(self, self.edges):
+            raise ValueError("the edges are not a valid cover of the graph")
+        # check there are no undirected cycles in the graph
+        if not nx.algorithms.tree.recognition.is_forest(self):
+            raise ValueError("the graph contains undirected cycles")
+        # check there are no disconnected segments
+        if not nx.algorithms.tree.recognition.is_tree(self):
+            raise ValueError("the graph contains disconnected segments")
```

Comment:
It's because we cannot directly test if an undirected graph contains cycles using the existed [algorithms](https://networkx.org/documentation/networkx-1.9/reference/algorithms.cycles.html).